Lab 33

*   Use our previous HBAT200 Dataset for practicing hyperparameter tuning.
*   Observe properly.
*   Have a documentation of your observations, make comparisons with the results we had from the same dataset before
*   Benchmark Linear Regression and Random Forest Regression.
*   Provide the following:
  - Comparison between the two algorithms.<br>
  - Answer and justify which model is better.

Explanation of Columns:

X1 - Customer Type

X2 - Industry Type

X3 - Firm Size

X4 - Region

X5 - Distribution System

X6 - Product Quality

X7 - E-Commerce

X8 - Technical Support

X9 - Complaint Resolution

X10 - Advertising

X11 - Product Line

X12 - Salesforce Image

X13 - Competitive Pricing X14 - Warranty & Claims

X15 - New Products

X16 - Order & Billing

X17 - Price Flexibility

X18 - Delivery Speed

X19 - Satisfaction

X20 - Likely to Recommend

X21 - Likely to Purchase

X22 - Purchase Level

X23 - Consider Strategic Alliance



#No Hyperparameter Tuning

In [ ]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, mean_squared_error, mean_absolute_error, r2_score
from sklearn.preprocessing import StandardScaler
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.feature_selection import SelectKBest, f_regression
from sklearn.metrics import mean_absolute_error, mean_absolute_percentage_error

# Create a DataFrame
df = pd.read_csv('HBAT-200-Data.csv')

df

ID  X1  X2  X3  X4  X5   X6   X7   X8   X9  ...  X14  X15  X16  X17  \
0      1   2   0   1   1   1  8.5  3.9  2.5  5.9  ...  4.7  4.3  5.0  5.1   
1      2   3   1   0   0   0  8.2  2.7  5.1  7.2  ...  5.5  4.0  3.9  4.3   
2      3   3   0   1   1   1  9.2  3.4  5.6  5.6  ...  6.2  4.6  5.4  4.0   
3      4   1   1   1   1   0  6.4  3.3  7.0  3.7  ...  7.0  3.6  4.3  4.1   
4      5   2   0   1   0   1  9.0  3.4  5.2  4.6  ...  6.1  4.5  4.5  3.5   
..   ...  ..  ..  ..  ..  ..  ...  ...  ...  ...  ...  ...  ...  ...  ...   
195  196   3   1   1   0   0  7.6  3.6  2.1  5.2  ...  4.3  6.8  4.4  3.9   
196  197   2   1   1   1   0  5.9  4.0  3.3  7.8  ...  5.2  6.6  6.0  7.5   
197  198   3   1   1   1   1  9.7  5.0  7.9  6.3  ...  6.6  5.2  4.4  3.5   
198  199   2   1   1   1   1  5.5  5.5  6.9  6.6  ...  7.0  5.3  5.9  6.7   
199  200   3   1   0   1   1  9.7  5.0  5.6  5.9  ...  6.9  1.2  4.3  3.5   

     X18  X19  X20  X21   X22  X23  
0    3.7  8.2  8.0  8.4  65.1    1  
1    4.9  5.7  6.5  7.5  67.1    0  
2    4.5  8.9  8.4  9.0  72.1    1  
3    3.0  4.8  6.0  7.2  40.1    0  
4    3.5  7.1  6.6  9.0  57.1    0  
..   ...  ...  ...  ...   ...  ...  
195  4.0  8.2  6.9  8.4  58.1    1  
196  5.4  7.9  7.8  8.6  62.1    1  
197  4.2  8.6  8.8  7.9  66.1    1  
198  4.9  8.2  7.1  7.6  59.1    1  
199  4.2  8.6  8.1  8.5  66.1    1  

[200 rows x 24 columns]

From Lab 23: Target(y) is X19, Selected Features(X) are: ['X6', 'X7', 'X9', 'X11', 'X12', 'X13', 'X17']

In [ ]:
X = df[['X6', 'X7', 'X9', 'X11', 'X12', 'X13', 'X17']]  # IVs
y = df['X19']

Split into training and testing sets

In [ ]:
# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=23)

Define and fit models, make predictions

In [ ]:
# Instantiate the regressors
lin_reg = LinearRegression()
rf_reg = RandomForestRegressor(random_state=0)

# Scaling for LinearRegression
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# np.abs(y_true - y_pred) computes absolute error btwn ea true valua and it's predicted value
# (y_true - y_pred) / y_true dividees absolute error by true value to get % error for each
# np.mean(np.abs((y_true - y_pred) / y_true))--np.abs() is applied again to ensure percentage errors are positive and
# np.mean() calculates the average of these absolute percentage errors and *100 reesults from fraction to %
def mean_absolute_percentage_error(y_true, y_pred):
    return np.mean(np.abs((y_true - y_pred) / y_true)) * 100

def fit_and_predict(model, X_train, y_train, X_test):
    model.fit(X_train, y_train)
    y_train_pred = model.predict(X_train)
    y_test_pred = model.predict(X_test)
    return y_train_pred, y_test_pred

# Fit and predict for both models
lin_reg_train_pred, lin_reg_test_pred = fit_and_predict(lin_reg, X_train_scaled, y_train, X_test_scaled)
rf_reg_train_pred, rf_reg_test_pred = fit_and_predict(rf_reg, X_train, y_train, X_test)


# Predictions for LinearRegression
print("LinearRegression - Training Predictions (First 10):")
print(lin_reg_train_pred[:10])  # Show the first 10 predictions

print("\nLinearRegression - Testing Predictions (First 10):")
print(lin_reg_test_pred[:10])

# Print 10 predictions for RandomForestRegressor
print("\nRandomForestRegressor - Training Predictions (First 10):")
print(rf_reg_train_pred[:10])

print("\nRandomForestRegressor - Testing Predictions (First 10):")
print(rf_reg_test_pred[:10])


LinearRegression - Training Predictions (First 10):
[7.32344442 7.09912427 5.89493417 8.68193511 5.67609725 8.46442132
 8.68327674 5.93866845 6.28976382 7.90781179]

LinearRegression - Testing Predictions (First 10):
[7.22972553 6.03412018 6.9364482  7.58144628 7.40903602 5.76354608
 5.44662876 8.78758237 8.24189454 6.20348838]

RandomForestRegressor - Training Predictions (First 10):
[7.868 7.779 5.975 8.885 5.464 8.509 8.755 6.303 6.313 8.585]

RandomForestRegressor - Testing Predictions (First 10):
[7.425 6.562 7.301 8.135 8.218 6.227 5.546 7.519 7.824 6.361]


Metrics/checking errors

In [ ]:
def calculate_metrics(y_true, y_pred):
    metrics = {
        'MSE': mean_squared_error(y_true, y_pred),
        'RMSE': np.sqrt(mean_squared_error(y_true, y_pred)),
        'MAE': mean_absolute_error(y_true, y_pred),
        'MAPE': mean_absolute_percentage_error(y_true, y_pred),
        'R²': r2_score(y_true, y_pred)
    }
    return metrics

def print_results(model_name, train_metrics, test_metrics):
    print(f"{model_name} - Regression:")
    print(f"Train - MSE: {train_metrics['MSE']:.4f}, RMSE: {train_metrics['RMSE']:.4f}, MAE: {train_metrics['MAE']:.4f}, MAPE: {train_metrics['MAPE']:.4f}%, R²: {train_metrics['R²']:.4f}")
    print(f"Test  - MSE: {test_metrics['MSE']:.4f}, RMSE: {test_metrics['RMSE']:.4f}, MAE: {test_metrics['MAE']:.4f}, MAPE: {test_metrics['MAPE']:.4f}%, R²: {test_metrics['R²']:.4f}")

# Calculate metrics for Linear Regression
lin_reg_train_metrics = calculate_metrics(y_train, lin_reg_train_pred)
lin_reg_test_metrics = calculate_metrics(y_test, lin_reg_test_pred)

# Calculate metrics for Random Forest
rf_reg_train_metrics = calculate_metrics(y_train, rf_reg_train_pred)
rf_reg_test_metrics = calculate_metrics(y_test, rf_reg_test_pred)

# Print results
print_results("LinearRegression", lin_reg_train_metrics, lin_reg_test_metrics)
print_results("RandomForestRegressor", rf_reg_train_metrics, rf_reg_test_metrics)



LinearRegression - Regression:
Train - MSE: 0.2549, RMSE: 0.5048, MAE: 0.3815, MAPE: 5.7125%, R²: 0.8499
Test  - MSE: 0.3980, RMSE: 0.6309, MAE: 0.4765, MAPE: 7.2917%, R²: 0.6879
RandomForestRegressor - Regression:
Train - MSE: 0.0260, RMSE: 0.1614, MAE: 0.1104, MAPE: 1.6886%, R²: 0.9847
Test  - MSE: 0.1438, RMSE: 0.3792, MAE: 0.2729, MAPE: 4.1036%, R²: 0.8872


Linear Regression:
- Performs well on training data with low MSE/RMSE, indicating it fits the training data well. MAE and MAPE suggests average prediction error is moderate.
- The testing data is slightly worse than the training data. The MSE, RMSE, MAE increased indicating it may not generalize as well to unseen data. The R^2 is lower, but okay, indicating the model explains less variance in the test data compared to the training data.

RandomForestRegression:
- Good performance on the training data. The R^2 is near 1.0 meaning it may explain almost all variance in the training data.
- The testing data is strong, but not as good as on the training data.  It may have some overfitting due to the MSE, RMSE and MAE are higher than the training data.

RandomForestRegressor may be the better performing model in terms of unseen data generalizing.

#Hyperparameter Tuning

##LinearRegression

Check default parameters

In [ ]:
# Instantiate Linear Regression model
lin_reg = LinearRegression()

# Get default hyperparameters for Linear Regression
default_params_lin_reg = lin_reg.get_params()
print("Default parameters for Linear Regression:", default_params_lin_reg)

Default parameters for Linear Regression: {'copy_X': True, 'fit_intercept': True, 'n_jobs': None, 'positive': False}


Best Hyperparameters

In [ ]:
param_grid = {
    'copy_X': [True, False],
    'fit_intercept': [True, False],
    'positive': [True, False],}

from sklearn.model_selection import GridSearchCV

# Create a GridSearchCV instance
grid_search = GridSearchCV(lin_reg, param_grid, cv=5, scoring='neg_mean_squared_error')

# Fit the model to perform the search (this will actually train the model)
grid_search.fit(X_train, y_train)

# Get the best hyperparameters from the search
best_params = grid_search.best_params_

# Print the best hyperparameters
print("\nBest Hyperparameters:")
for param, value in best_params.items():
    print(f"{param}: {value}")


Best Hyperparameters:
copy_X: True
fit_intercept: True
positive: False


Redefine the model with the best parameters, training, and predicting.

In [ ]:
# Get the best model from the grid search
best_model = grid_search.best_estimator_

# Train the best model on the training data
best_model.fit(X_train, y_train)

y_pred_train = best_model.predict(X_train)

# Now, you can use the trained model for predictions on new data (X_test)
y_pred = best_model.predict(X_test)

In [ ]:
# Function to calculate and return metrics
def calculate_metrics(y_true, y_pred):
    metrics = {
        'MSE': mean_squared_error(y_true, y_pred),
        'RMSE': np.sqrt(mean_squared_error(y_true, y_pred)),
        'MAE': mean_absolute_error(y_true, y_pred),
        'MAPE': np.mean(np.abs((y_true - y_pred) / y_true)) * 100,
        'R²': r2_score(y_true, y_pred)
    }
    return metrics

# Get metrics for training data
train_metrics = calculate_metrics(y_train, y_pred_train)

# Get metrics for test data
test_metrics = calculate_metrics(y_test, y_pred)

# Print results in the desired format
print("LinearRegression - Regression:")
print(f"Train - MSE: {train_metrics['MSE']:.4f}, RMSE: {train_metrics['RMSE']:.4f}, MAE: {train_metrics['MAE']:.4f}, MAPE: {train_metrics['MAPE']:.4f}%, R²: {train_metrics['R²']:.4f}")
print(f"Test  - MSE: {test_metrics['MSE']:.4f}, RMSE: {test_metrics['RMSE']:.4f}, MAE: {test_metrics['MAE']:.4f}, MAPE: {test_metrics['MAPE']:.4f}%, R²: {test_metrics['R²']:.4f}")


LinearRegression - Regression:
Train - MSE: 0.2549, RMSE: 0.5048, MAE: 0.3815, MAPE: 5.7125%, R²: 0.8499
Test  - MSE: 0.3980, RMSE: 0.6309, MAE: 0.4765, MAPE: 7.2917%, R²: 0.6879


The results are identical to non hyperparameter tuning. There was no change in model performance. Reason is because the parameters used are the same as the default which are already near optimal for this model. Any tweaking of parameters will not improve performance.

##RandomForestRegression

Check default parameters

In [ ]:
# Instantiate the RandomForestRegressor model # do not use all or computer will crash
rf_reg = RandomForestRegressor(random_state=0)

# Get the default hyperparameters
default_params = rf_reg.get_params()
print("Default Hyperparameters for RandomForestRegressor:")
default_params

Default Hyperparameters for RandomForestRegressor:


{'bootstrap': True,
 'ccp_alpha': 0.0,
 'criterion': 'squared_error',
 'max_depth': None,
 'max_features': 1.0,
 'max_leaf_nodes': None,
 'max_samples': None,
 'min_impurity_decrease': 0.0,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'n_estimators': 100,
 'n_jobs': None,
 'oob_score': False,
 'random_state': 0,
 'verbose': 0,
 'warm_start': False}

Best Hyperparameters

In [ ]:
param_grid = {
    'n_estimators': [50, 100, 200],
    'max_depth': [None, 10, 20],
    'min_samples_split': [2, 5, 10]}

from sklearn.model_selection import GridSearchCV

# Create a GridSearchCV instance
grid_search = GridSearchCV(model, param_grid, cv=5)

# Fit the model to perform the search (this will actually train the model with cross-validation)
grid_search.fit(X_train, y_train)

# Get the best hyperparameters from the search
best_params = grid_search.best_params_

# Print the best hyperparameters
print("\nBest Hyperparameters:")
for param, value in best_params.items():
    print(f"{param}: {value}")


Best Hyperparameters:
max_depth: 10
min_samples_split: 2
n_estimators: 100


Redefine the model with the best parameters, training, and predicting.

In [ ]:
# Get the best model from the grid search
best_model = grid_search.best_estimator_

# Train the best model on the training data
best_model.fit(X_train, y_train)

y_pred_train = best_model.predict(X_train)

# Now, you can use the trained model for predictions on new data (X_test)
y_pred = best_model.predict(X_test)

In [ ]:
# Redefine with the best parameters
best_rf_reg = RandomForestRegressor(
    n_estimators=100,
    max_depth=10,
    min_samples_split=2,
    random_state=0
)

# Train the model on the training data
best_rf_reg.fit(X_train, y_train)

# Make predictions on the training data
y_pred_train = best_rf_reg.predict(X_train)

# Make predictions on the testing data
y_pred_test = best_rf_reg.predict(X_test)

# Function to calculate and return metrics
def calculate_metrics(y_true, y_pred):
    metrics = {
        'MSE': mean_squared_error(y_true, y_pred),
        'RMSE': np.sqrt(mean_squared_error(y_true, y_pred)),
        'MAE': mean_absolute_error(y_true, y_pred),
        'MAPE': np.mean(np.abs((y_true - y_pred) / y_true)) * 100,
        'R²': r2_score(y_true, y_pred)
    }
    return metrics

# Get metrics for training data
train_metrics = calculate_metrics(y_train, y_pred_train)

# Get metrics for testing data
test_metrics = calculate_metrics(y_test, y_pred_test)

# Print results in the desired format
print("RandomForestRegressor - Regression:")
print(f"Train - MSE: {train_metrics['MSE']:.4f}, RMSE: {train_metrics['RMSE']:.4f}, MAE: {train_metrics['MAE']:.4f}, MAPE: {train_metrics['MAPE']:.4f}%, R²: {train_metrics['R²']:.4f}")
print(f"Test  - MSE: {test_metrics['MSE']:.4f}, RMSE: {test_metrics['RMSE']:.4f}, MAE: {test_metrics['MAE']:.4f}, MAPE: {test_metrics['MAPE']:.4f}%, R²: {test_metrics['R²']:.4f}")


RandomForestRegressor - Regression:
Train - MSE: 0.0261, RMSE: 0.1616, MAE: 0.1103, MAPE: 1.6881%, R²: 0.9846
Test  - MSE: 0.1431, RMSE: 0.3783, MAE: 0.2718, MAPE: 4.0925%, R²: 0.8878


The tuning has very small metric improvements on the training. The testing metrics show a slight improvement. The hyperparamter adjustments are not noteworthy.

#Summary

# Regression Model Comparison

| Metric | LinearRegression | | | | RandomForestRegressor | | | |
|--------|------------------|------------------|------------------|------------------|------------------|------------------|------------------|------------------|
|        | Without Tuning   |                  | With Tuning      |                  | Without Tuning   |                  | With Tuning      |                  |
|        | Train            | Test             | Train            | Test             | Train            | Test             | Train            | Test             |
| R^2    | 0.8499           | 0.6879           | 0.8499           | 0.6879           | 0.9847           | 0.8872           | 0.9846           | 0.8878           |
| MAPE   | 5.7125%          | 7.2917%          | 5.7125%          | 7.2917%          | 1.6886%          | 4.1036%          | 1.6881%          | 4.0925%          |
| MAE    | 0.3815           | 0.4765           | 0.3815           | 0.4765           | 0.1104           | 0.2729           | 0.1103           | 0.2718           |
| RMSE   | 0.5048           | 0.6309           | 0.5048           | 0.6309           | 0.1614           | 0.3792           | 0.1616           | 0.3783           |
| MSE    | 0.2549           | 0.3980           | 0.2549           | 0.3980           | 0.0260           | 0.1438           | 0.0261           | 0.1431           |

<br>
R^2 = accuracy<br>
MAPE = error<br>

**Before/After tuning**

- Before tuning
  - Random Forest Regressor metrics illustrated a good fit, but not the best performance for testing set.
  - Both may be overfitted (Train has higher R^2 than test, train has lower MAPE than test and confirmed by other metrics)

- After tuning
  - Random Forest Regressor showed slight improvements though not monumental.
  - Linear Regression metrics remained identical, indicating the default model was already very well tuned.<br>

**Benchmarking**
- Linear Regression
  - Simpler model
  - Consistent metrics indicating stability with default parameters

- Random Forest Regressor
  - More complex model
  - Slight improvements after tuning, showing tuning can have marginal gains
<br>

**Comparison**
- Metrics
  - Random Forest performs better in terms of R^2, MSE, RMSE. The slight improvement after tuning indicates the model benefits from tuning adjustments, although minor.
- Complexity
  - Random Forest is more complex with more parameters to tune and can fit more complex patterns in the data.
  - Linear Regression is simpler and may not capture complex patterns.
<br>

**Which Model is better**

- Here, Random Forest Regressor is better regarding the performance metrics (MSE, RMSE, R^2).
- The slight improvement after tuning for Random Forest Regressor suggests the model benefits from hyperparameter adjustments, although slightly.

